### ISSUE TITLE
Board Review: Search Batching

### ISSUE THREAD

**Issue by tg-msft** [Azure/azure-sdk#2241]
## Contacts and Timeline

* Responsible service team: **Azure Search v-team**
* Main contacts: **adpsearch**
* Expected code complete date: **Already in Preview**
* Expected release date: **February**

## About the Service 

* Link to documentation introducing/describing the service: **https://azure.microsoft.com/en-us/services/search/**
* Link to the service REST APIs: **https://docs.microsoft.com/en-us/rest/api/searchservice/**
* Link to GitHub issue for previous review sessions, if applicable: **https://github.com/Azure/azure-sdk/issues/1637 and https://github.com/Azure/azure-sdk/issues/1769**

## About the client library

* Name of the client library: **Azure.Search.Documents**
* Languages for this review: **C#, Java, Python, TypeScript**

### .NET

* APIView Link: **[.NET](https://apiview.dev/Assemblies/Review/f291bdd4488149748454e2e2fa18a821?diffRevisionId=df5833f95a854d249bd4f9bda3d24692&doc=False)**
* Link to Champion Scenarios/Quickstart samples: **[Champion Scenarios](https://gist.github.com/tg-msft/478b5db739f1296f53ea96acd51bd027)**

### Java

* APIView Link: **[Java](https://apiview.dev/Assemblies/Review/a6b0f1ad79c147b4b553fa47d2113a35/e3703bca31a649858a617cf175f96503?doc=False)** _(Is there a better ApiView with a diff @alzimmermsft?)_
* Link to Champion Scenarios/Quickstart samples: **COMING SOON**

### Python

* APIView Link: **[Python](https://apiview.dev/Assemblies/Review/16dd4cb824fd4382a62456b1002b3de1?diffRevisionId=ef5df2c702b24d788a54937d696402b3&doc=False)**
* Link to Champion Scenarios/Quickstart samples: **COMING SOON**
  
### TypeScript

* APIView Link: **[TypeScript](https://apiview.dev/Assemblies/Review/0477e6efbd3641f79548178dec485e89)**  _(Is there a better ApiView with a diff @sarangan12?)_
* Link to Champion Scenarios/Quickstart samples: **COMING SOON**

We'll also follow up with additional details on the semantics of Search Batching before the meeting.

**Comment by lilyjma** [comment]
scheduled for 1/13

**Comment by tg-msft** [comment]
# .NET's Implementation

This is a high level description of the algorithm used for batching in .NET.

## Concepts

The major moving pieces:

- `SearchClient`: the regular client class with an `IndexDocuments` method that
   performs simple batching.  This is tied to a single Search index.  Our fancy
   batching will use one of these instances to handle all communication with the
   service.  _This has already GA'ed._

- `IndexDocumentsAction<T>`: The model type used for simple and fancy batching
  APIs representing an `Upload`, `Merge`, `MergeOrUpload`, or `Delete` of a
  document in a Search index.  _This has already GA'ed._

- `SearchIndexingBufferedSender`: This is our fancy batch "object."  It's not
  quite a client, not quite a model.  It's full of methods for adding index
  actions as fast as possible that mirror the convenience methods on
  `SearchClient`.  This type will be thread-safe so it can be slammed with
  index actions.

- Batch Publisher: We're going to have a "many producers, single publisher"
  model of interaction.  The publisher will not be publicly exposed but handle
  all the conceptual heavy lifting.

## Obtaining a SearchIndexingBufferedSender

- `SearchClient.CreateIndexingBufferedSender<T>(SearchIndexingBufferedSenderOptions<T>)`
    - Instantiates a new `SearchIndexingBufferedSender` tied to the current client.
    - The sender knows about the type of documents in its index
        - Could also be used with `SearchDocument` for unknown document types
    - The only way of obtaining a `SearchIndexingBufferedSender` (i.e., no .ctor)
    - `virtual` as the entry point for mocking fancy batching scenarios

## Configuring SearchIndexingBufferedSender

- `public class SearchIndexingBufferedSenderOptions<T> { ... }`

    - The generic parameter `T` is the type of documents the
    `SearchIndexingBufferedSender` will manage.  At one point this type had to
    be generic for the events, but now it's only for the `KeyFieldAccessor`
    which could be simplified to take `object` if we wanted to remove the type
    parameter here.

    - `bool AutoFlush { get; set; } = true;`

    - `TimeSpan? AutoFlushInterval { get; set; } = TimeSpan.FromSeconds(60);`

    - `CancellationToken FlushCancellationToken  { get; set; }`
    This is the (optional) cancellation token used by the async publisher.

    - `Func<T, string> KeyFieldAccessor { get; set; }`  Used to get the key name
    of a document in case we need to associate errors with their inputs.  We'll
    use this first if provided.

    - `private int? InitialBatchActionCount  { get; set; }`  The initial number
    of actions to group into a batch when tuning the behavior of the sender.
    The default value will be 512 if unset.  The current service maximum is 32000.

    - `public int MaxRetries { get; set; } = 3;`  
    - `public TimeSpan RetryDelay { get; set; } = TimeSpan.FromSeconds(0.8);`
    - `public TimeSpan MaxRetryDelay { get; set; } = TimeSpan.FromMinutes(1);`

        Describes the retry settings for failed documents.  The MaxRetries is
        the number of times an individual document can fail across batches.  The
        RetryDelay and MaxRetryDelay correspond to delay added before batches
        when we're being throttled.

## Using SearchIndexingBufferedSender

- `public class SearchIndexingBufferedSender<T>`
    - Add Indexing Actions
        - `void DeleteDocuments(IEnumerable<T>, CancellationToken)`
        - `Task DeleteDocumentsAsync(IEnumerable<T>, CancellationToken)`
        - `void IndexDocuments(IndexDocumentsBatch<T>, CancellationToken)`
        - `Task IndexDocumentsAsync(IndexDocumentsBatch<T>, CancellationToken)`

            Note these are using a batch which we already have conveniences for
            building up.

        - `void MergeDocuments(IEnumerable<T>, CancellationToken)`
        - `Task MergeDocumentsAsync(IEnumerable<T>, CancellationToken)`
        - `void MergeOrUploadDocuments(IEnumerable, CancellationToken)`
        - `Task MergeOrUploadDocumentsAsync(IEnumerable<T>, CancellationToken)`
        - `void UploadDocuments(IEnumerable<T>, CancellationToken)`
        - `Task UploadDocumentsAsync(IEnumerable<T>, CancellationToken)`

            We're not returning any data from these operations.  They may submit
            something immediately or we might be waiting for a flush.

            The provided cancellation token will be used for any requests sent, any
            notifications raised, etc.  If a flush happens when the timer elapses,
            then we'll use the `FlushCancellationToken` instead.  A  customer could
            reuse a token or link them together for fine grained control.

            We're also not including the `IndexDocumentsOptions` parameter because
            that currently only serves to configure exception behavior which we're
            doing exclusively through events.

        - Delete by key

            Add customization for "delete with just a key" scenarios for dynamic
            `SearchDocument`.  We're using extension methods in .NET.

            - `public static class SearchExtensions`
                - `static void DeleteDocuments(this SearchIndexingBufferedSender<SearchDocument>, string, IEnumerable<string>, CancellationToken)`
                - `static Task DeleteDocumentsAsync(this SearchIndexingBufferedSender<SearchDocument>, string, IEnumerable<string>, CancellationToken)`

    - `void Flush(CancellationToken)` and `Task FlushAsync(CancellationToken)`
        will manually flush any pending operations.  This could result in
        multiple requests.  If a request is already in progress, whatever's left
        will still be flushed on returning.  This will wait for the flush to
        complete.

    - Notifications will be raised as events.
        - `event Func<IndexDocumentsAction<T>, CancellationToken, Task> ActionAddedAsync;`
        - `event Func<IndexDocumentsAction<T>, CancellationToken, Task> ActionSentAsync;`
        - `event Func<IndexDocumentsAction<T>, IndexingResult, CancellationToken, Task> ActionCompletedAsync;`
        - `event Func<IndexDocumentsAction<T>, IndexingResult, Exception, CancellationToken, Task> ActionFailedAsync;`

            Either the `IndexingResult` or `Exception` may be null for failures
            depending on the type of failure and how it was surfaced.

            We'll raise either ActionCompleted or ActionFailed events.  You'd
            need to handle both to implement checkpointing.

            We will not throw any exceptions from the pipeline.  If a failure
            is triggered from a timer, there's no good way for a customer to
            catch it.  Everything will go through the notification events
            instead.  If we find the need for request level failures, we could
            add a `public event Action<Exception, CancellationToken> BatchFailed;`
            or similar.

            All of the events are async only becuase they're being called from
            the async only Publisher.  This may change depending on other
            discussions around .NET's events in Track 2.  When raising an async
            event, we'll collect all the returned tasks from each handler and
            `Task.WaitAll` to ensure they've all run to completion.

    - Cleaning up with `Dispose`

        - `void Dispose()` will call the publisher's `Dispose`, which does a
        blocking flush.

        - `Task DisposeAsync()` will call the publisher's `DisposeAsync`, which
        does a blocking flush.

        - We need the publisher to block on these particular calls because the
        application might exit as soon as `Dispose` returns.

        - We'll set `bool _disposed = true` and throw `ObjectDisposedException`
        if anyone tries to call any of our methods after disposing.  We'll also
        suppress the finalizer since no new work can take place.

        - We'll add a finalizer/destructor to check for any remaining work if
        we haven't been properly disposed.  In .NET, we can't safely throw an
        exception because it would tear down the process.  Instead we'll throw
        an exception that we immediately catch and swallow ourselves which still
        shows up in logs.  We'll also add a custom logging event for this
        specific scenario if anyone's listening to our event source.

## Publishing

Most of the heavy lifting will live in an `internal class SearchIndexingPublisher<T>`.
and a base `Publisher<T>` that might be a good candidate for internal shared
source when we get a second use case.

- Uses `System.Threading.Channels` to for a CSP approach where the Sender
  produces messages that the Publisher consumes and processes.  This allows us
  to have a many senders (even though it's the single object, it can be shared
  across threads) and single publisher model.  The only messages are `Publish`
  a number of actions and `Flush` the pending messages..

- "Public" API of the internal Publisher

    These are what `SearchIndexingBufferedSender<T>` will use.

    - `void AddDocumentsAsync(IEnumerable<IndexDocumentsAction<T>>, CancellationToken)`
    - `Task FlushAsync(CancellationToken)`
    - `void Dispose(CancellationToken)`
    - `Task DisposeAsync(CancellationToken)`

        They largely encapsulate sending messages through the channel that the
        message loop will process.

    - `public int IndexingActionsCount` This is the number of pending
    actions that haven't been submitted.  It's primarily intended to indicate
    whether the customer needs to apply backpressure higher up the stack.
    We're keeping it effectively private for now while we think about how
    backpressure is exposed.

    We'll also use properties/event handlers on the publisher as the backing
    store for the `SearchClient`, various options, and events.  This is primarily
    to avoid the publisher needing to reach back into the sender.

- Synchronization

    We need to make sure `SearchIndexingBufferedSender<T>` is thread-safe per
    our guidelines even though it has mutable state in the form of pending
    actions.

    The Search service has undefined semantics around parallel requests updating
    the same document.  We're also far more likely to get throttled if we submit
    simultaneous requests.  We're going to restrict things so that only one
    thread at a time can send requests.  All of the pending queues and other
    mutable data structures are manipulated from the single thread running our
    message loop.

- Validate
    - Ensure we have a key field accessor
    - This needs to happen on the calling thread so we can raise exceptions

- Getting a Key Field Accessor
    - If we haven't set a `_keyFieldAccessor`
        - Default to `Options.KeyFieldAccessor`
        - If that's null, use the logic in `FieldBuilder` to find the property
        in `T` with `IsKey=true` set and use its `Getter` as the accessor
        - If that didn't work (because `T` wasn't used with `FieldBuilder`),
        we'll get a `SearchIndexClient` from our `SearchClient` and make a
        `GetIndex` call to the service and compare properties to fields.  We
        will not cover cases like custom naming policies.
        - Set `_keyFieldAccessor` to the value we found
        - If `_keyFieldAccessor == null` throw an exception
    - Return `_keyFieldAccessor`

- Adding actions
    - The `AddDocumentsAsync` method will add everything to the pending
    actions queue.  It will create an object that includes the action, the
    action's key (using the `_keyFieldAccessor`), the document for the action,
    and a retry count.  We could consider tracking intermediate failures if we
    want to expose those in the future.
    - Raise the `ActionAddedAsync` event
    - If `AutoFlush` is `true`
        - If we have enough documents to submit a batch
            - Call `PublishAsync(flush: false)`
        - If there are pending actions
            - Start the idle timer to AutoFlush

- Flushing
    - The `Flush`/`Async` methods will call `Publish(flush: true)`.  After the
    publish is finished, we'll notify anyone waiting on our `TaskCompletionSource`
    that flushing has completed.

- Disposing
    - The `Dispose`/`DisposeAsync` methods will call `Flush()` and wait for the
    messages to be sent.  We'll also set a `_disposed` flag on the publisher so
    calling `Dispose` from every sender thread works safely.  It also shuts down
    the `Channel`, waits for the message loop to finish processing any last
    messages, and disposes the timer.

- `Publish(bool flush)`
    - Disable any existing timer
    - If `flush || (_pendingSize + _retrySize >= BatchSize)`
        - `FillBatchFromQueue` with retries first then pending
        - `SubmitBatchAsync(batch)`
    - `SubmitBatch(batch)`
    - If there's enough for another batch, keep looping through `Publish`

- Filling Batches
    - A batch will consist of `InitialBatchActionCount` wrapped actions
    - `FillFromQueue(batch, _retryActions) && FillFromQueue(batch, _pendingActions)`
    - return `batch`
    - `bool FillFromQueue(batch, queue)`
        - `while (queue.TryPeek(out action))`
            - If `batch.Length > InitialBatchActionCount`
                - Dequeue and add to the batch
            - else return `true`
        - Return `false`
    - This prefers retrying actions before processing new actions.
    - This also prefers keeping actions ordered.
    - In the future, it would be nice to also do this based on the batch binary
      size.

- `SubmitBatchAsync(List<Action<T>> batch)`
    - Optionally add a manual retry delay if the next submission is too close
    to the previous submission and we're being throttled.
    - Raise the ActionSentAsync event
    - Send the filled batch as a request using our `SearchClient`.
    - If we get a 413 Payload Too Large
        - Split the batch in half.
        - Add the second half to the retry actions (without incrementing the retry count).
        - Recurse on the first half.
    - If we get a 200 or 207
        - Associate errors with their actions from the batch
        - For each error
            - If the `RetryCount` is less than the max, increment the count and
            add it to the `_retryActions` queue
            - Else raise the `ActionFailed` event and log the failure
        - For each success
            - Raise the `ActionCompleted` event
        - If any of the errors were 503s
            - Add an exponential retry delay with hitter based on the number of
            attempts (following `RetryPolicy`'s implementation)
            - call `SubmitBatch(batch, retryAttempt + 1)`

- Associating Errors
    - Given a batch and an indexing response, we'll need to associate errors
    with actions.  This is nontrivial.
    - Walk through the items matching failures to their keys
        - If there are multiple items with the same key, we're not in a good spot.
        - For now we will have to rely on the service ordering (i.e., `Kth`
        response for a key corresponds to the `Kth` requested item)
        - We'll stable sort by key and walk both lists together for efficiency.

- Using the timer
    - Use `System.Threading.Timer` in .NET even though it only uses a sync
    callback.
    - Keep track of the `_timer` instance so it can be disabled/disposed.
    - Send's a `Flush` message to through the channel (ignoring the returned
    `Task`) and uses the `FlushCancellationToken`

## Future work

- Logging

    We need to add a custom `EventSource` for logging so folks can get deep
    insight into exactly what's happening and how they can tune the performance.
    I think we'll start with events for each notification, an event for unfinished
    work, an event for splitting a batch, etc.  We should also include data on
    the current batch size and possibly summary statistics.

- PID controllers for autoscaling along:
    - Batch size
    - Submission delay (using retry with jitter today)
    - Backpressure delay for producers

**Comment by lilyjma** [comment]
[Recording](https://msit.microsoftstream.com/video/e449a1ff-0400-9887-aa77-f1eb55fa9603)[MS INTERNAL ONLY]
